### PR TITLE
Type annotated binders

### PR DESCRIPTION
**Pull request by yiming-fang**
This PR adds type annotations/holes to all binders in lets, lambdas, and patterns, which get filled during the elaboration pass of type inference.

### CODE REVIEW

**Comment by j-hui** [comment]
This is passing all tests locally. Thank you so much @yiming-fang and @leoqiao18 for helping me port this! 

Since the changes are quite large, I'm just going to go ahead and format all the files we touched. In particular, I think the typechecker and `IR.IR` need reformatting since I never did those post-Brittany.

**Comment by j-hui** [comment]
Oh right, @yiming-fang  is still working on getting lambda lifting working with recursive functions. Since that's orthogonal to this PR, maybe you could merge that in on a separate PR?

**Comment by yiming-fang** [comment]
> Oh right, @yiming-fang is still working on getting lambda lifting working with recursive functions (which should also close #111). Since that's orthogonal to this PR, maybe you could merge that in on a separate PR?

Sure, I can open a separate PR for lambda lifting mutually recursive lets.

File: src/IR/Constraint/Constrain/Expression.hs [hunk 1]
```diff
@@ -1,50 +1,54 @@
...
+-- let annsVarPairs = map (uncarryAttachment . fst) binderDefs
+--     -- vars = map snd annsVarPairs
+--     exprTyps = map (TVarN . snd . uncarryAttachment . snd) binderDefs
+--     forBinders ((ann, var), exprTyp) =
+--       exists [var] <$> do
+--         Ann.withAnnotations ann (TVarN var) \varExp ->
+--           return $ CEqual varExp exprTyp
+-- constraints <- mapM forBinders (zip annsVarPairs exprTyps)
+-- return $ CAnd (consDefs : constraints)
+
```
Comment (leoqiao18) [inline src/IR/Constraint/Constrain/Expression.hs:51]: I forgot to remove this commented-out block of code when I committed my fix. Could you help me clean this up @yiming-fang

Comment (leoqiao18) [reply src/IR/Constraint/Constrain/Expression.hs:51]: Other than this, I have no other suggestions. This PR is looking good!

Comment (yiming-fang) [reply src/IR/Constraint/Constrain/Expression.hs]: > I forgot to remove this commented-out block of code when I committed my fix. Could you help me clean this up @yiming-fang

Done!

**Pull request merged**